File: 2015/05/javascript/main.mjs
import fs from "node:fs";
import readline from "node:readline";

// --- Constants ---
const INPUT_FILE = "../input/input.txt";
const LIST_1 = ['a', 'e', 'i', 'o', 'u'];
const LIST_2 = ["ab", "cd", "pq", "xy"];

// --- Rules ---
function rule_1_1(s) {
	let count = 0;
	for (const c of Array.from(s)) {
		if (LIST_1.includes(c)) count++;
		if (count == 3) return true;
	}
	return false;
}

function rule_1_2(s) {
	const a = Array.from(s);
	for (let i = 0; i < a.length - 1; i++) {
		if (a[i] == a[i + 1]) return true;
	}
	return false;
}

function rule_1_3(s) {
	for (const c of LIST_2) {
		if (s.includes(c)) return false;
	}
	return true;
}

function rule_2_1(s) {
	const a = Array.from(s);
	for (let i = 0; i < a.length - 2; i++) {
		for (let j = i + 2; j < a.length - 1; j++) {
			if (a[i] == a[j] && a[i + 1] == a[j + 1]) return true;
		}
	}
	return false;
}

function rule_2_2(s) {
	const a = Array.from(s);
	for (let i = 0; i < a.length - 2; i++) {
		if (a[i] == a[i + 2]) return true;
	}
	return false;
}

// --- Variables ---
let count1 = 0;
let count2 = 0;

// --- Read and parse the input file ---
const stream = fs.createReadStream(INPUT_FILE);
const reader = readline.createInterface({ input: stream, crlDelay: Infinity });
reader.on("line", line => {
	if (rule_1_1(line) && rule_1_2(line) && rule_1_3(line)) count1++;
	if (rule_2_1(line) && rule_2_2(line)) count2++;
});
await new Promise(res => reader.once("close", res));

// --- Puzzle 1 ---
console.log("1. Nice Strings: ", count1.toLocaleString());

// --- Puzzle 2 ---
console.log("2. Nice Strings: ", count2.toLocaleString());

File: 2015/05/python/main.py
# --- Constants --
INPUT_FILE = "../input/input.txt"
LIST_1 = ['a', 'e', 'i', 'o', 'u']
LIST_2 = ["ab", "cd", "pq", "xy"]

# --- Functions ---
def rule_1_1(s):
    ls = [*s]
    count = 0
    for c in ls:
        try:
            LIST_1.index(c)
            count += 1
            if count == 3: return True
        except ValueError:
            pass
    return False

def rule_1_2(s):
    ls = [*s]
    for i in range(0, len(ls) - 1):
        if ls[i] == ls[i + 1]: return True
    return False

def rule_1_3(s):
    for c in LIST_2:
        if s.find(c) != -1: return False
    return True

def rule_2_1(s):
    ls = [*s]
    for i in range(0, len(ls) - 2):
        for j in range(i + 2, len(ls) - 1):
            if ls[i] == ls[j] and ls[i + 1] == ls[j + 1]: return True
    return False

def rule_2_2(s):
    ls = [*s]
    for i in range(0, len(ls) - 2):
            if ls[i] == ls[i + 2]: return True
    return False

# --- Variables ---
count1 = 0
count2 = 0

# --- Read and Parse the input file --
file = open(INPUT_FILE, "r")
while True:
    line = file.readline()
    if not line: break
    line = line.strip()

    # --- Puzzle 1 ---
    if rule_1_1(line) and rule_1_2(line) and rule_1_3(line): count1 += 1

    # --- Puzzle 1 ---
    if rule_2_1(line) and rule_2_2(line): count2 += 1

file.close()

# --- Results ---
print(f"1. Nice Strings: {count1:,d}")
print(f"2. Nice Strings: {count2:,d}")
